parser/object_parse/die_parse/item_name.rs

```diff
@@ -18,13 +29,19 @@ pub fn item_name<R>(
 where
     R: gimli::Reader,
 {
-    if let Some(attr) = die.attr_value(gimli::DW_AT_name)? {
-        let s = dwarf.attr_string(unit, attr)?;
-        Ok(Some(
-            s.to_string()? // This `to_string()` creates a `Result<Cow<'_, str>, _>`.
-                .to_string(), // This `to_string()` creates the String we return.
-        ))
-    } else {
-        Ok(None)
-    }
+    die.attr_value(gimli::DW_AT_linkage_name)
+        .transpose()
+        .or_else(|| die.attr_value(gimli::DW_AT_name).transpose())
+        .transpose()?
+        .map(
+            |attr: gimli::read::AttributeValue<R>| -> Result<String, traits::Error> {
+                Ok(
+                    dwarf
+                        .attr_string(unit, attr)?
+                        .to_string()? // This `to_string()` creates a `Result<Cow<'_, str>, _>`.
+                        .to_string(), // This `to_string()` creates the String we return.
+                )
+            },
+        )
+        .transpose()
```

Comment:
This is the meat of the PR, and it's admittedly dense. Apologies!

The alternative was outlining the contents of the old `if let` block into a helper function, which was too vertically noisy for my taste, or (_hiss_) duplicating the code into a new `else if let`.

It _is_ a little clever*, but the `transpose` calls let us lazily call `die.attr_value(gimli::DW_AT_name)` only in the case that we have an `Ok(None)`, while still bailing early if we've encountered some bad data. The final `transpose` is to let us use `?`'s in our mapping closure, and bubble the result back up so we're matching the type signature.

* in situations like this, I mean "clever" with the pejorative connotation.

Reply:
This is all subjective, so do what you prefer, but I would do:
```rust
let mut attr = die.attr_value(gimli::DW_AT_linkage_name)?;
if attr.is_none() {
    attr = die.attr_value(gimli::DW_AT_name)?;
}
if let Some(attr) = attr {
    //unchanged
   ...
}
```

Reply:
Thanks for providing some feedback! This led me to think about it a little more, and I think I found a nice way to do this with a `match` statement [here](https://github.com/rustwasm/twiggy/pull/402/files#diff-a5729e8c41bb69cc717e50cab8e2da98R33). No mutability! :tada: